v03_pipeline/lib/methods/sex_check.py

```diff
@@ -0,0 +1,215 @@
+    :param aaf_threshold: Alternate allele frequency threshold for `hl.impute_sex`. Default is 0.05
+    :param af_field: Name of field containing allele frequency information. Default is 'info.AF'
+    :param call_rate_threshold: Minimum required call rate. Default is 0.25
+    :return Table with imputed sex annotations, and the fstat plot.
+    """
+
+    valid_contigs = {*reference_genome.autosomes, *reference_genome.sex_chromosomes}
+    if normalization_contig not in valid_contigs:
+        msg = f'Contig: {normalization_contig} is invalid for this reference genome'
+        raise ValueError(msg)
+
+    # Filter to SNVs and biallelics
+    # NB: We should already have filtered biallelics, but just
+    mt = mt.filter_rows(hl.is_snp(mt.alleles[0], mt.alleles[1]))
+
+    # Filter to PASS variants only (variants with empty or missing filter set)
+    # TODO: Make this an optional argument before moving to gnomad_methods
+    mt = mt.filter_rows(
+        hl.is_missing(mt.filters) | (mt.filters.length() == 0),
+        keep=True,
+    )
+
+    impute_sex_ht = hl.impute_sex(
+        mt.GT,
+        male_threshold=xy_fstat_threshold,
+        female_threshold=xx_fstat_threshold,
+        aaf_threshold=aaf_threshold,
+    )
+    ht = mt.annotate_cols(**impute_sex_ht[mt.col_key]).cols()
+
+    annotations = [
+        *IMPUTE_SEX_ANNOTATIONS,
+    ]
+    if not use_chrY_cov:
+        ht = ht.annotate(
+            sex=(
+                hl.case()
+                .when(hl.is_missing(ht.is_female), Ploidy.AMBIGUOUS.value)
+                .when(ht.is_female, Ploidy.FEMALE.value)
+                .default(Ploidy.MALE.value)
+            ),
+        )
+        return ht.select(*annotations)
+
+    annotations = [
+        *IMPUTE_SEX_ANNOTATIONS,
+        f'{normalization_contig}_mean_dp',
+        'chrY_mean_dp',
+        'normalized_y_coverage',
+        'chrX_mean_dp',
+        'normalized_x_coverage',
+    ]
+    for contig in [
+        normalization_contig,
+        hl.get_reference(reference_genome.value).y_contigs[0],
+        hl.get_reference(reference_genome.value).x_contigs[0],
+    ]:
+        contig_ht = get_contig_cov(
+            mt,
+            reference_genome,
+            contig,
+            call_rate_threshold,
+            af_field,
+        )
+        ht = ht.annotate(**contig_ht[ht.s])
+    ht = ht.annotate(
+        normalized_y_coverage=hl.or_missing(
+            ht[f'{normalization_contig}_mean_dp'] > 0,
+            ht.chrY_mean_dp / ht[f'{normalization_contig}_mean_dp'],
+        ),
+        normalized_x_coverage=hl.or_missing(
+            ht[f'{normalization_contig}_mean_dp'] > 0,
+            ht.chrX_mean_dp / ht[f'{normalization_contig}_mean_dp'],
+        ),
+    )
+    ht = ht.annotate(
+        sex=(
+            hl.case()
+            .when(hl.is_missing(ht.is_female), Ploidy.AMBIGUOUS.value)
+            .when(
```

Comment:
this is mostly shared with the `not use_chrY_cov` case above, it might be better to have a helper function to run all the shared code, with an optional argument to determine whether or not to add the one `when` clause thats different, or to reorganize the function so instead of having a branching return with duplicated logic and just have a single path to the return that handles that conditions

Reply:
I cleaned this up a bit.